Add docs on task-specific buffering using multithreading (#48542)
JuliaLang/julia: doc/src/manual/multi-threading.md

Note that [`Threads.@threads`](@ref) does not have an optional reduction parameter like [`@distributed`](@ref).

### Using `@threads` without data races

Taking the example of a naive sum

```julia-repl
julia> function sum_single(a)
           s = 0
           for i in a
               s += i
           end
           s
       end
sum_single (generic function with 1 method)

julia> sum_single(1:1_000_000)
500000500000
```

Simply adding `@threads` exposes a data race with multiple threads reading and writing `s` at the same time.
```julia-repl
julia> function sum_multi_bad(a)
           s = 0
           Threads.@threads for i in a
               s += i
           end
           s
       end
sum_multi_bad (generic function with 1 method)

julia> sum_multi_bad(1:1_000_000)
70140554652
```

Note that the result is not `500000500000` as it should be, and will most likely change each evaluation.

To fix this, buffers that are specific to the task may be used to segment the sum into chunks that are race-free.

```julia-repl

> julia> function sum_multi_good(a)

Reviewer: I'm not an expert on multi-threaded parallelism, so I might just be wrong.

But this doesn't seem like it'll be particularly performant as the values of the `WeakKeyDict` won't be contiguous in memory, and so the reduction at the needs to gather these from likely different parts of the heap => slow reduction. In fact, when I tried running something like this on an example of mine it was incredibly slow; probably partially due to what I just mentioned, and partially due to the slowness of `WeakKeyDict` (and using something like `Dict` doesn't actually seem work).

An alternative is `Atomic` as mentioned below, or, even better, you can do a `Vector{Atomic{T}}` of some `buffer_length`. In the latter scenario you can then just randomly pick an index for each `Task`, and act on the corresponding `Atomic{T}` atomically, e.g. `atomic_add!(buffer[rand(1:buffer_size)], i)`. If the work within each of the tasks is fairly uniform, then this random picking of index to add to should result in fairly evenly congestion for the different buffers.

I'm sure there are much, much better ways of doing this, but the above is a quick-and-easy way to implement a much faster version of this kind of map-reduce approach using a buffer.

>            buffers = WeakKeyDict{Task,Int}()

Reviewer 2: This shouldn't be weak, otherwise it might be empty when you go to sum it

           Threads.@threads for i in a
               b = get(buffers, current_task(), 0)
               b += i
               buffers[current_task()] = b
           end
           s = 0
           for b in values(buffers)
               s += b
           end
           s
       end
sum_multi_good (generic function with 1 method)

julia> sum_multi_good(1:1_000_000)
500000500000
```

> Note that we do not use buffers based on the `threadid()` i.e. `buffers = zeros(Threads.nthreads())` because tasks can

Reviewer 2: Also because nthreads is soft-deprecated (no warning, but may return incorrect answers now)

Author: I didn't add a note on this. I don't understand whether the docs should or shouldn't recommend `nthreads()`

actually change thread at yield points, known as task migration (docs??), which would introduce data races for such a
buffering approach.

> Another option, which would not be performant for a naive sum like this, is the use of atomic operations on variables shared

Reviewer 2: This would probably would be much faster for this particular example however, so it might be confusing in this context here

Author: Updatedd

Note that [`Threads.@threads`](@ref) does not have an optional reduction parameter like [`@distributed`](@ref).

### Using `@threads` without data races

Taking the example of a naive sum

```julia-repl
julia> function sum_single(a)
           s = 0
           for i in a
               s += i
           end
           s
       end
sum_single (generic function with 1 method)

julia> sum_single(1:1_000_000)
500000500000
```

Simply adding `@threads` exposes a data race with multiple threads reading and writing `s` at the same time.
```julia-repl
julia> function sum_multi_bad(a)
           s = 0
           Threads.@threads for i in a
               s += i
           end
           s
       end
sum_multi_bad (generic function with 1 method)

julia> sum_multi_bad(1:1_000_000)
70140554652
```

Note that the result is not `500000500000` as it should be, and will most likely change each evaluation.

To fix this, buffers that are specific to the task may be used to segment the sum into chunks that are race-free.

```julia-repl

> julia> function sum_multi_good(a)
>            buffers = WeakKeyDict{Task,Int}()
>            Threads.@threads for i in a
>                b = get(buffers, current_task(), 0)
>                b += i
>                buffers[current_task()] = b
>            end
>            s = 0
>            for b in values(buffers)
>                s += b
>            end
>            s
>        end

Reviewer 3: ```suggestion
julia> function sum_multi_good(a)
           chunks = Iterators.partition(a, length(a) ÷ Threads.nthreads())
           tasks = map(chunks) do chunk
               Threads.@spawn begin
                   b = zero(eltype(chunk))
                   for i ∈ chunk
                       b += i
                   end
                   return b
               end
           end
           s = 0
           buffers = fetch.(tasks)
           for b in buffers
               s += b
           end
           s
       end
```

Note that [`Threads.@threads`](@ref) does not have an optional reduction parameter like [`@distributed`](@ref).

### Using `@threads` without data races

Taking the example of a naive sum

```julia-repl
julia> function sum_single(a)
           s = 0
           for i in a
               s += i
           end
           s
       end
sum_single (generic function with 1 method)

julia> sum_single(1:1_000_000)
500000500000
```

Simply adding `@threads` exposes a data race with multiple threads reading and writing `s` at the same time.
```julia-repl
julia> function sum_multi_bad(a)
           s = 0
           Threads.@threads for i in a
               s += i
           end
           s
       end
sum_multi_bad (generic function with 1 method)

julia> sum_multi_bad(1:1_000_000)
70140554652
```

Note that the result is not `500000500000` as it should be, and will most likely change each evaluation.

To fix this, buffers that are specific to the task may be used to segment the sum into chunks that are race-free.

```julia-repl
julia> function sum_multi_good(a)
           chunks = Iterators.partition(a, length(a) ÷ Threads.nthreads())
           tasks = map(chunks) do chunk

>                Threads.@spawn begin
>                    b = zero(eltype(chunk))
>                    for i ∈ chunk
>                        b += i
>                    end
>                    return b
>                end

Reviewer 3: Maybe this is a bit nicer?
```suggestion
               Threads.@spawn sum_single(chunk)
```

Author: Yeah. Updated with a new note

           end

>            s = 0
>            buffers = fetch.(tasks)
>            for b in buffers
>                s += b
>            end
>            s

Reviewer 3: ```suggestion
           sum_single(fetch.(tasks))
```

       end
sum_multi_good (generic function with 1 method)

julia> sum_multi_good(1:1_000_000)
500000500000
```

Note buffers should not be managed based on the `threadid()` i.e. `buffers = zeros(Threads.nthreads())` because tasks can
actually change thread at yield points, known as [task migration](@ref man-task-migration), which introduces potential for

> data races for such a buffering approach.

Reviewer 3: I'm not very clear on the lingo here, but I think the main concern is actually `yield`ing, not migration -- or do we consider `yield` a form of migration even if the `threadid` doesn't change?

Author: Yeah. Should be fixed now
